core/lib/vm_executor/src/oneshot/mod.rs

```diff
@@ -275,30 +373,90 @@ impl<S: ReadStorage> VmSandbox<S> {
         }
     }
 
-    pub(super) fn apply<T, F>(mut self, apply_fn: F) -> T
-    where
-        F: FnOnce(&mut LegacyVmInstance<S, HistoryDisabled>, Transaction) -> T,
-    {
+    /// This method is blocking.
+    fn execute_in_vm<T>(
+        mut self,
+        action: impl FnOnce(&mut Vm<StorageWithOverrides<S>>, Transaction) -> T,
+    ) -> T {
+        Self::setup_storage(
+            &mut self.storage,
+            &self.execution_args,
+            self.env.current_block,
+        );
+
+        let protocol_version = self.env.system.version;
+        let mode = self.env.system.execution_mode;
+        if self.execution_args.adjust_pubdata_price {
+            self.env.l1_batch.fee_input = adjust_pubdata_price_for_tx(
+                self.env.l1_batch.fee_input,
+                self.execution_args.transaction.gas_per_pubdata_byte_limit(),
+                self.env.l1_batch.enforced_base_fee.map(U256::from),
+                protocol_version.into(),
+            );
+        };
+
+        let transaction = self.execution_args.transaction;
         let tx_id = format!(
             "{:?}-{}",
-            self.transaction.initiator_account(),
-            self.transaction.nonce().unwrap_or(Nonce(0))
+            transaction.initiator_account(),
+            transaction.nonce().unwrap_or(Nonce(0))
         );
 
+        let storage_view = StorageView::new(self.storage).to_rc_ptr();
+        let mut vm = match self.fast_vm_mode {
+            FastVmMode::Old => Vm::Legacy(LegacyVmInstance::new_with_specific_version(
+                self.env.l1_batch,
+                self.env.system,
+                storage_view.clone(),
+                protocol_version.into_api_vm_version(),
+            )),
+            FastVmMode::New => Vm::Fast(FastVmInstance::fast(
+                self.env.l1_batch,
+                self.env.system,
+                storage_view.clone(),
+            )),
+            FastVmMode::Shadow => {
+                let mut vm =
+                    ShadowVm::new(self.env.l1_batch, self.env.system, storage_view.clone());
+                if !self.panic_on_divergence {
+                    let transaction = format!("{:?}", transaction);
+                    let handler = DivergenceHandler::new(move |errors, _| {
+                        tracing::error!(transaction, ?mode, "{errors}");
+                    });
```

Comment:
Not sure what the best option for the divergence handler would look like. We probably don't want to panic on divergence, but maybe we want to store a VM dump somewhere?

Reply:
It does, maybe we can just dump it to local file with some safety measures against spamming files, e.g. limit it to 10 dump files and overwrite/stop dumping when limit is reached?

Reply:
Will we be able to get these files from e.g. mainnet pods (where divergences are empirically most likely to happen)? AFAIU, this would require K8s privileges mere mortals don't have (like executing commands in a pod).